lms/static/scripts/frontend_apps/components/test/BookSelector-test.js

```diff
@@ -0,0 +1,224 @@
+      const wrapper = renderBookSelector();
+      updateURL(wrapper, 'foo');
+
+      assert.calledOnce(fakeBookIDFromURL);
+
+      updateURL(wrapper, '');
+
+      assert.calledOnce(fakeBookIDFromURL);
+    });
+
+    it('shows an error if entered URL format is invalid', () => {
+      fakeBookIDFromURL.returns(null);
+
+      const wrapper = renderBookSelector();
+      updateURL(wrapper, 'foo');
+
+      const errorMessage = wrapper.find('[data-testid="error-message"]');
+
+      assert.isTrue(errorMessage.exists());
+      assert.include(
+        errorMessage.text(),
+        "That doesn't look like a VitalSource book link"
+      );
+      assert.isTrue(errorMessage.find('SvgIcon[name="cancel"]').exists());
+    });
+  });
+
+  context('correctly-formatted URL entered', () => {
+    it('fetches book metadata associated with the entered URL', async () => {
+      const onSelectBook = sinon.stub();
+      const wrapper = renderBookSelector({ onSelectBook });
+      updateURL(wrapper, 'a valid URL');
+
+      await waitFor(() => onSelectBook.called);
+
+      assert.calledWith(fakeVitalSourceService.fetchBook, 'BOOKSHELF_TUTORIAL');
+      assert.calledOnce(onSelectBook);
+      assert.calledWith(onSelectBook, bookData.bookList[0]);
+    });
+
+    it('clears any existing book metadata before loading and selecting new book', async () => {
+      const onSelectBook = sinon.stub();
+      const selectedBook = bookData.bookList[0];
+      const wrapper = renderBookSelector({ onSelectBook, selectedBook });
+
+      updateURL(wrapper, 'a valid URL');
+
+      await waitFor(() => onSelectBook.called);
+
+      assert.equal(onSelectBook.callCount, 2);
+      assert.equal(
+        onSelectBook.getCall(0).args[0],
+        null,
+        'clears the pre-existing selected book'
+      );
+      assert.equal(
+        onSelectBook.getCall(1).args[0],
+        bookData.bookList[0],
+        'selects the newly-fetched book'
+      );
+    });
+
+    it('disables input and associated button while fetching', async () => {
+      const onSelectBook = sinon.stub();
+      const wrapper = renderBookSelector({ onSelectBook });
+      updateURL(wrapper, 'a valid URL');
+
+      const input = getInput(wrapper);
+      const iconButton = wrapper.find('IconButton button[title="Find book"]');
+
+      assert.isTrue(input.props().disabled);
+      assert.isTrue(iconButton.props().disabled);
+
+      await waitForElement(wrapper, 'IconButton[disabled=false]');
+
+      assert.isFalse(getInput(wrapper).props().disabled);
+    });
+
+    it('shows cover thumbnail in loading state while fetching', async () => {
+      const onSelectBook = sinon.stub();
+      const wrapper = renderBookSelector({ onSelectBook });
+
+      assert.isFalse(wrapper.find('Thumbnail img').exists());
+
+      updateURL(wrapper, 'a valid URL');
+
+      assert.isTrue(wrapper.find('Thumbnail').props().isLoading);
+
+      return await waitForElement(wrapper, 'Thumbnail[isLoading=false]');
```

Comment:
```suggestion
      await waitForElement(wrapper, 'Thumbnail[isLoading=false]');
```

The `return` isn't needed. If the `waitForElement` promise rejects then `await` will throw an error and the test will fail.

Reply:
OK, yep. This is an artifact of some flailing I was experiencing with the combination of `async` needs in these tests. I'm currently ruminating on my experience to see if there's any way we could smooth the way for future devs. It wasn't nightmarish or anything, but it took me about 3-4 hours to write tests for this PR branch, which felt a little long.